gen/goav2gen/definition.go

```diff
@@ -0,0 +1,255 @@
+		}
+		queryName += "[" + name + "]"
+	}
+	if p.Type == "Array" {
+		queryName += "[]"
+	}
+	ap := &gen.ActionParam{
+		Name:        name,
+		QueryName:   queryName,
+		Description: cleanDescription(p.Description),
+		VarName:     toVarName(name),
+		Location:    gen.PayloadParam,
+		Mandatory:   required,
+		NonBlank:    (required || p.Pattern != "") && (p.Default != nil),
+		Regexp:      p.Pattern,
+		ValidValues: p.Enum,
+	}
+	if min, ok := p.Minimum.(float64); ok {
+		ap.Min = int(min)
+	}
+	if max, ok := p.Maximum.(float64); ok {
+		ap.Max = int(max)
+	}
+
+	if p.IsRef() {
+		ap.Type = a.typeForRef(ec, p.GetRef())
+		return ap
+	}
+	switch p.Type {
+	case "array":
+		elemT := a.typeForRef(ec.WithTrail(name+"[]"), p.Items)
+		apElem := &gen.ActionParam{
+			Location: gen.PayloadParam,
+			Type:     elemT,
+		}
+		ap.Type = &gen.ArrayDataType{
+			ElemType: apElem,
+		}
+	case "object":
+		if a.Doc.Ref(p.Schema) != nil {
+			ap.Type = a.typeForRef(ec.WithTrail(name), p.Schema)
+		} else {
+			dt := gen.EnumerableDataType(0)
+			ap.Type = &dt
+		}
+	default:
+		if p.Format == "date-time" {
+			a.api.NeedTime = true
+			t := gen.BasicDataType("*time.Time")
+			ap.Type = &t
+		} else if t := a.overrideAttr(ec, name); t != nil {
+			ap.Type = t
+		} else {
+			ap.Type = basicType(p.Type)
+		}
+	}
+
+	return ap
+}
+
+func (a *APIAnalyzer) overrideAttr(ec EvalCtx, name string) gen.DataType {
+	if override, ok := a.AttrOverrides[name]; ok {
+		switch override {
+		case "sourcefile":
+			return new(gen.UploadDataType)
+		}
+		return basicType(override)
+	}
+	return nil
+}
+
+// typeForRef converts a "$ref" reference in the swagger definition to a concrete type, usually an object.
+func (a *APIAnalyzer) typeForRef(ec EvalCtx, r Ref) gen.DataType {
+	switch r.Type() {
+	case "array":
+		fail("References to arrays not implemented yet")
+	case "object":
+		def := a.Doc.Ref(r)
+		if def == nil {
+			fail("No ref for %+v", r)
+		}
+		dt := a.AnalyzeDefinition(ec, def, r.ID())
+		if objDT, ok := dt.(*gen.ObjectDataType); ok {
+			a.addType(ec, objDT, r)
+		}
+		return dt
+	}
+	return basicType(r.Type())
+
+}
+
+// addType conditionally adds a new type, trying its best to avoid type collisions
+func (a *APIAnalyzer) addType(ec EvalCtx, dt *gen.ObjectDataType, r Ref) {
+	a.api.NeedJSON = true
+	if a.refByType[dt.TypeName] == r.ID() {
+		return
+	}
+
+	if other, ok := a.api.Types[dt.TypeName]; ok {
+		if !ec.IsResult {
+			// If its an input parameter, fix the collision by amending this types name
+			dt.TypeName += "Param"
+			if a.refByType[dt.TypeName] == r.ID() {
+				return
+			}
+		}
+		oldFields := []string{}
+		newFields := []string{}
+		for _, f := range other.Fields {
+			oldFields = append(oldFields, f.Name)
+		}
+		for _, f := range dt.Fields {
+			newFields = append(newFields, f.Name)
+		}
+		use := "Old"
+		if len(newFields) > len(oldFields) {
+			use = "New"
+		}
+		if strings.Join(oldFields, ",") != strings.Join(newFields, ",") {
```

Comment:
What is this stuff? What is it comparing?

Reply:
Ah so this is the downside of swagger 2 vs swagger 3. For swagger 2 if you have multiple things a given type like an applied policy or user they all have to create their own definition for it in the swagger "definitions" section. So they'll be X copies of identical definition for a User struct lets say. This tries to collapse those things down into one return type struct. Apparently swagger 3 reorgs things a bit so they'll all point to the same thing if they are the same thing as there won't be any guess work.